File: fhir4ds/cql/duckdb/macros/math.py
"""
CQL Math functions as DuckDB SQL macros.

Tier 1 implementation - zero Python overhead.
These macros are inlined at query planning time.
"""

from decimal import ROUND_HALF_UP, Decimal, InvalidOperation, localcontext
from typing import TYPE_CHECKING

import duckdb

if TYPE_CHECKING:
    import duckdb


def _cql_divide(a, b):
    """CQL §16.4 divide: exact Decimal division at implementation scale 8.

    DuckDB's native ``/`` always promotes DECIMAL operands to DOUBLE,
    producing binary floating-point artifacts (e.g. ``9.9 / 3.0`` ->
    3.3000000000000003). CQL defines divide as *Decimal* division limited
    to the implementation precision/scale (28/8 here), so the result must
    be quantized half-up to 8 fractional digits, not a DOUBLE. Returns
    NULL for NULL operands, a zero divisor, non-numeric operands, or a
    result outside the representable Decimal range (spec: "If the result
    of the division cannot be represented ... the result is null").
    Operands arrive as VARCHAR (DuckDB casts DECIMAL/ints exactly; DOUBLE
    via its shortest round-trip representation).
    """
    if a is None or b is None:
        return None
    try:
        da = a if isinstance(a, Decimal) else Decimal(str(a).strip())
        db = b if isinstance(b, Decimal) else Decimal(str(b).strip())
    except (ValueError, InvalidOperation, ArithmeticError):
        return None
    if db == 0:
        return None
    try:
        with localcontext() as ctx:
            ctx.prec = 60
            quotient = da / db
            result = quotient.quantize(Decimal("1.00000000"), rounding=ROUND_HALF_UP)
    except (InvalidOperation, ArithmeticError):
        return None
    if result.copy_abs() >= Decimal(10) ** 28:
        # Exceeds the implementation Decimal range (28 digits, scale 8).
        return None
    return result


def registerMathMacros(con: "duckdb.DuckDBPyConnection") -> None:
    """
    Register math function macros (Tier 1).

    Registers native DuckDB SQL macros for CQL math functions:
    - Abs, Ceiling, Floor, Round, RoundTo, Sqrt, Exp, Ln, Log, Power
    - Truncate, Sign, Mod, Div, CQLMessage

    All functions have zero Python overhead.

    Note: Uses 'system.' prefix to reference built-in functions and avoid
    infinite recursion when macro name matches the function name.
    """
    # Direct mappings to native DuckDB functions (use system. prefix to avoid recursion)
    con.execute("CREATE OR REPLACE MACRO Abs(x) AS TRY(system.abs(x))")
    # CQL 1.5 Appendix B Ceiling/Floor: "If the result of the operation
    # cannot be represented as an Integer, the result is null."
    # TRY_CAST to INTEGER nulls out-of-Integer-extent results (e.g.
    # Ceiling(3147483647.05), Floor(2147483648.2)) and NULL inputs.
    con.execute(
        "CREATE OR REPLACE MACRO Ceiling(x) AS TRY_CAST(system.ceiling(x) AS INTEGER)"
    )
    con.execute(
        "CREATE OR REPLACE MACRO Floor(x) AS TRY_CAST(system.floor(x) AS INTEGER)"
    )

    # Round - CQL Appendix B: traditional rounding. Half ties round away from
    # zero, so -0.5 becomes -1 and -1.5 becomes -2.
    # CQL-11 HISTORIAN QA-001: use system.floor/system.ceil, NOT the bare
    # FLOOR/CEIL names — Floor(x) is overridden above as the CQL Floor macro
    # (TRY_CAST to INTEGER), which nulls any rounded magnitude above the
    # Integer range (e.g. Round(3147483647.05) -> NULL instead of
    # 3147483647). Round returns a Decimal, so only Decimal-unrepresentable
    # results (|x| >= 10^30) are null: TRY_CAST to DECIMAL(38, 8).
    con.execute(
        "CREATE OR REPLACE MACRO Round(x) AS "
        "CASE WHEN x IS NULL THEN NULL ELSE TRY_CAST("
        "CASE WHEN CAST(x AS DOUBLE) >= 0 "
        "THEN system.floor(CAST(x AS DOUBLE) + 0.5) "
        "ELSE system.ceil(CAST(x AS DOUBLE) - 0.5) END "
        "AS DECIMAL(38, 8)) END"
    )
    con.execute(
        "CREATE OR REPLACE MACRO RoundTo(x, prec) AS "
        "CASE WHEN x IS NULL THEN NULL ELSE TRY_CAST(("
        "CASE WHEN CAST(x AS DOUBLE) * POWER(10, COALESCE(prec, 0)) >= 0 "
        "THEN system.floor(CAST(x AS DOUBLE) * POWER(10, COALESCE(prec, 0)) + 0.5) "
        "ELSE system.ceil(CAST(x AS DOUBLE) * POWER(10, COALESCE(prec, 0)) - 0.5) END"
        ") / POWER(10, COALESCE(prec, 0)) AS DECIMAL(38, 8)) END"
    )

    # Other math functions
    con.execute("CREATE OR REPLACE MACRO Sqrt(x) AS TRY(system.sqrt(x))")
    # CQL v1.5.3 §16.6 Exp: "If the result of the operation cannot be
    # represented, the result is null." Reinforced by section header:
    # "operations that cause arithmetic overflow or underflow ... will
    # result in null, rather than a run-time error." So return NULL on
    # overflow (e.g. Exp(710), Exp(1000)), not a runtime error.
    # CQL signature Exp(argument Decimal) Decimal — the result is a
    # Decimal at the implementation scale (8), not a DOUBLE.
    con.execute(
        "CREATE OR REPLACE MACRO Exp(x) AS "
        "CASE "
        "WHEN x IS NULL THEN NULL "
        "WHEN isfinite(TRY(system.exp(CAST(x AS DOUBLE)))) "
        "THEN TRY_CAST(system.exp(x) AS DECIMAL(38, 8)) "
        "ELSE NULL END"
    )
    # CQL v1.5.3 §16.12 Ln: "If the result of the operation cannot be
    # represented, the result is null." Ln(0) is -infinity (cannot be
    # represented) and Ln(negative) is undefined; both return NULL per
    # spec, not runtime errors.
    # CQL signature Ln(argument Decimal) Decimal — Decimal result.
    con.execute(
        "CREATE OR REPLACE MACRO Ln(x) AS "
        "CASE "
        "WHEN x IS NULL THEN NULL "
        "WHEN TRY_CAST(x AS DOUBLE) = 0 THEN NULL "
        "WHEN TRY_CAST(x AS DOUBLE) < 0 THEN NULL "
        "WHEN isfinite(TRY(system.ln(CAST(x AS DOUBLE)))) "
        "THEN TRY_CAST(system.ln(x) AS DECIMAL(38, 8)) "
        "ELSE NULL END"
    )
    # CQL Log is the arbitrary-base two-argument operator. Natural log is Ln.
    # CQL signature Log(argument Decimal, base Decimal) Decimal.
    con.execute(
        "CREATE OR REPLACE MACRO Log(x, base) AS "
        "CASE WHEN isfinite(TRY(system.ln(CAST(x AS DOUBLE)) / system.ln(CAST(base AS DOUBLE)))) "
        "THEN TRY_CAST(system.ln(x) / system.ln(base) AS DECIMAL(38, 8)) ELSE NULL END"
    )
    con.execute(
        "CREATE OR REPLACE MACRO LogBase(x, base) AS "
        "CASE WHEN isfinite(TRY(system.ln(CAST(x AS DOUBLE)) / system.ln(CAST(base AS DOUBLE)))) "
        "THEN system.ln(x) / system.ln(base) ELSE NULL END"
    )

    con.execute(
        "CREATE OR REPLACE MACRO Power(x, y) AS "
        "CASE "
        "WHEN x IS NULL OR y IS NULL THEN NULL "
        "WHEN isfinite(TRY(system.pow(TRY_CAST(x AS DOUBLE), TRY_CAST(y AS DOUBLE)))) "
        "THEN system.pow(TRY_CAST(x AS DOUBLE), TRY_CAST(y AS DOUBLE)) "
        "ELSE NULL END"
    )
    con.execute("CREATE MACRO IF NOT EXISTS Truncate(x) AS TRY_CAST(system.trunc(x) AS INTEGER)")
    con.execute("CREATE MACRO IF NOT EXISTS Sign(x) AS system.sign(x)")

    # Modulo and integer division
    con.execute("CREATE MACRO IF NOT EXISTS Mod(x, y) AS x % y")
    con.execute("CREATE MACRO IF NOT EXISTS Div(x, y) AS system.trunc(x / NULLIF(y, 0))")

    # CQL §16.4 divide (`/`): exact Decimal division. DuckDB `/` promotes
    # DECIMAL operands to DOUBLE (9.9 / 3.0 -> 3.3000000000000003), which
    # violates the Decimal result type and scale-8 quantization CQL
    # requires. Scalar-UDF (Tier 3) because exact base-10 division is not
    # expressible with DuckDB's DOUBLE-only decimal divide. Registered as
    # cqlDivide and referenced directly by the translator's divide lowering.
    try:
        con.create_function(
            "cqlDivide",
            _cql_divide,
            parameters=["VARCHAR", "VARCHAR"],
            return_type=duckdb.decimal_type(38, 8),
            null_handling="special",
        )
    except (duckdb.CatalogException, duckdb.InvalidInputException):
        # Already registered on this connection (repeated extension setup).
        pass

    # CQL Appendix B Errors and Messaging: Message returns source unchanged,
    # except true-condition Error severity raises a runtime error.
    con.execute(
        """
        CREATE OR REPLACE MACRO CQLMessage(source, condition, code, severity, message) AS
        CASE
            WHEN COALESCE(condition, FALSE) AND lower(CAST(severity AS VARCHAR)) = 'error'
            THEN error(COALESCE(CAST(code AS VARCHAR), '') || ': ' || COALESCE(CAST(message AS VARCHAR), ''))
            ELSE source
        END
        """
    )


__all__ = ["registerMathMacros"]

File: extensions/cql/src/cql/math.cpp
#include "cql/math.hpp"
#include <cmath>
#include <cstdlib>
#include <sstream>
#include <cstring>
#include <cfloat>
#include <stdexcept>

namespace cql {

// =====================================================================
// Helper: parse string to double
// =====================================================================
static bool parse_double(const std::string &s, double &out) {
	if (s.empty()) return false;
	char *end = NULL;
	out = std::strtod(s.c_str(), &end);
	return (end != s.c_str() && *end == '\0' && !std::isinf(out) && !std::isnan(out));
}

static std::string format_result(double val) {
	// Integer result
	if (val == std::floor(val) && std::fabs(val) < 1e15 && !std::isinf(val)) {
		std::ostringstream oss;
		oss << static_cast<long long>(val);
		return oss.str();
	}
	std::ostringstream oss;
	oss.precision(15);
	oss << val;
	return oss.str();
}

// =====================================================================
// Abs
// =====================================================================
Optional<std::string> math_abs(const std::string &x) {
	double val;
	if (!parse_double(x, val)) return NullOpt<std::string>();
	return Optional<std::string>(format_result(std::fabs(val)));
}

// =====================================================================
// Ceiling
// =====================================================================
Optional<std::string> math_ceiling(const std::string &x) {
	double val;
	if (!parse_double(x, val)) return NullOpt<std::string>();
	// CQL 1.5 Appendix B: Ceiling(Decimal) returns Integer; results that
	// cannot be represented as an Integer are null.
	double result = std::ceil(val);
	if (std::isnan(result) || std::isinf(result) ||
	    result < -2147483648.0 || result > 2147483647.0) {
		return NullOpt<std::string>();
	}
	return Optional<std::string>(format_result(result));
}

// =====================================================================
// Floor
// =====================================================================
Optional<std::string> math_floor(const std::string &x) {
	double val;
	if (!parse_double(x, val)) return NullOpt<std::string>();
	// CQL 1.5 Appendix B: Floor(Decimal) returns Integer; results that
	// cannot be represented as an Integer are null.
	double result = std::floor(val);
	if (std::isnan(result) || std::isinf(result) ||
	    result < -2147483648.0 || result > 2147483647.0) {
		return NullOpt<std::string>();
	}
	return Optional<std::string>(format_result(result));
}

// =====================================================================
// Exp
// =====================================================================
Optional<std::string> math_exp(const std::string &x) {
	double val;
	if (!parse_double(x, val)) return NullOpt<std::string>();
	double result = std::exp(val);
	// CQL v1.5.3 §16.6 Exp: "If the result of the operation cannot be
	// represented, the result is null." Reinforced by section header:
	// "operations that cause arithmetic overflow or underflow ... will
	// result in null, rather than a run-time error." So we return NULL
	// (NullOpt) when the result overflows to infinity (e.g. Exp(710),
	// Exp(1000), Exp(1e5)), instead of raising a runtime error.
	if (std::isinf(result)) {
		return NullOpt<std::string>();
	}
	return Optional<std::string>(format_result(result));
}

// =====================================================================
// Ln — natural logarithm
// =====================================================================
Optional<std::string> math_ln(const std::string &x) {
	double val;
	if (!parse_double(x, val)) return NullOpt<std::string>();
	// CQL v1.5.3 §16.12 Ln: "If the result of the operation cannot be
	// represented, the result is null." Ln(0) is -infinity and cannot be
	// represented, so return NULL rather than raising a runtime error.
	if (val == 0) {
		return NullOpt<std::string>();
	}
	if (val < 0) return NullOpt<std::string>();
	return Optional<std::string>(format_result(std::log(val)));
}

// =====================================================================
// Log(x, base)
// =====================================================================
Optional<std::string> math_log(const std::string &x, const std::string &base) {
	double xval, bval;
	if (!parse_double(x, xval) || !parse_double(base, bval)) return NullOpt<std::string>();
	if (xval <= 0 || bval <= 0 || bval == 1.0) return NullOpt<std::string>();
	return Optional<std::string>(format_result(std::log(xval) / std::log(bval)));
}

// =====================================================================
// Power
// =====================================================================
Optional<std::string> math_power(const std::string &x, const std::string &exp) {
	double xval, eval;
	if (!parse_double(x, xval) || !parse_double(exp, eval)) return NullOpt<std::string>();
	double result = std::pow(xval, eval);
	if (std::isinf(result) || std::isnan(result)) return NullOpt<std::string>();
	if (result != 0.0 && std::fabs(result) < 1e-4) {
		// Sub-scale magnitudes must be emitted in fixed notation at the
		// implementation scale (8): '%.15g'-style scientific text is
		// rounded UP to 1E-8 by DuckDB's VARCHAR->DECIMAL(38,8) cast,
		// so Power(2, -30) returned 0.00000001 instead of the
		// correctly quantized 0.00000000.
		char buf[64];
		std::snprintf(buf, sizeof(buf), "%.8f", result);
		return Optional<std::string>(std::string(buf));
	}
	return Optional<std::string>(format_result(result));
}

// =====================================================================
// Round — CQL half-up rounding (not banker's rounding)
// =====================================================================
Optional<std::string> math_round(const std::string &x, const std::string &precision) {
	double val;
	int prec;
	if (!parse_double(x, val)) return NullOpt<std::string>();
	char *end = NULL;
	prec = static_cast<int>(std::strtol(precision.c_str(), &end, 10));
	if (end == precision.c_str()) prec = 0;

	// CQL traditional rounding moves negative half ties away from zero:
	// the reference explicitly gives Round(-0.5) = -1.
	double multiplier = std::pow(10.0, prec);
	double shifted = val * multiplier;
	if (shifted >= 0.0) {
		shifted = std::floor(shifted + 0.5);
	} else {
		shifted = std::ceil(shifted - 0.5);
	}
	double result = shifted / multiplier;

	if (prec <= 0) {
		return Optional<std::string>(format_result(result));
	}
	// Format with exact decimal places
	std::ostringstream oss;
	oss.precision(prec);
	oss << std::fixed << result;
	return Optional<std::string>(oss.str());
}

// =====================================================================
// Sqrt
// =====================================================================
Optional<std::string> math_sqrt(const std::string &x) {
	double val;
	if (!parse_double(x, val)) return NullOpt<std::string>();
	if (val < 0) return NullOpt<std::string>();
	return Optional<std::string>(format_result(std::sqrt(val)));
}

// =====================================================================
// Truncate — integer part toward zero
// =====================================================================
Optional<std::string> math_truncate(const std::string &x) {
	double val;
	if (!parse_double(x, val)) return NullOpt<std::string>();
	double result = (val >= 0) ? std::floor(val) : std::ceil(val);
	return Optional<std::string>(format_result(result));
}

// =====================================================================
// cqlDivide — exact Decimal division at the implementation scale (8).
//
// CQL §16.4 divide: the result is a Decimal quantized half-up (ties away
// from zero) to 8 fractional digits. DuckDB's native `/` promotes DECIMAL
// operands to DOUBLE (9.9 / 3.0 -> 3.3000000000000003), so the exact
// quotient must be computed with base-10 long division over digit
// strings. Mirrors the Python authority `_cql_divide`
// (fhir4ds/cql/duckdb/macros/math.py): NULL for NULL operands (handled
// by the UDF wrapper), non-numeric operands, a zero divisor, or a result
// whose magnitude reaches 10^28 (28 integer digits after rounding).
//
// Rounding-equivalence note: the Python authority divides at prec=60 and
// then quantizes ROUND_HALF_UP to scale 8. For every result Python
// returns (integer part <= 28 digits, so quantize never overflows the
// 60-digit context), the scale-8 guard digit at fractional position 9
// lies inside the preserved 60-significant-digit window, and half-even
// carries beyond position 60 can only reach position 9 through a run of
// 9s that HALF_UP rounds up anyway. Direct HALF_UP rounding of the true
// quotient at scale 8 is therefore byte-identical to the two-step
// Python result.
// =====================================================================
namespace {

const size_t CQL_DIVIDE_MAX_DIGITS = 5000;
const long long CQL_DIVIDE_MAX_EXP = 5000;

int compare_digit_strings(const std::string &a, const std::string &b) {
	if (a.size() != b.size()) return a.size() < b.size() ? -1 : 1;
	int c = a.compare(b);
	return c < 0 ? -1 : (c > 0 ? 1 : 0);
}

std::string subtract_digit_strings(const std::string &a, const std::string &b) {
	// a >= b, both without leading zeros ("0" allowed)
	std::string res(a.size(), '0');
	int borrow = 0;
	for (int i = static_cast<int>(a.size()) - 1, j = static_cast<int>(b.size()) - 1; i >= 0; --i, --j) {
		int da = a[i] - '0';
		int db = (j >= 0 ? b[j] - '0' : 0) + borrow;
		if (da < db) {
			da += 10;
			borrow = 1;
		} else {
			borrow = 0;
		}
		res[i] = static_cast<char>('0' + da - db);
	}
	size_t nz = res.find_first_not_of('0');
	return nz == std::string::npos ? "0" : res.substr(nz);
}

struct DecimalParts {
	int sign;
	std::string digits; // no leading zeros; "0" for the value 0
	long long exp10;    // value = sign * digits * 10^exp10
};

bool parse_cql_decimal_text(const std::string &raw, DecimalParts &out) {
	size_t b = 0, e = raw.size();
	auto is_ws = [](char c) {
		return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\v' || c == '\f';
	};
	while (b < e && is_ws(raw[b])) b++;
	while (e > b && is_ws(raw[e - 1])) e--;
	if (b >= e) return false;
	size_t p = b;
	out.sign = 1;
	if (raw[p] == '+' || raw[p] == '-') {
		if (raw[p] == '-') out.sign = -1;
		p++;
	}
	std::string int_digits, frac_digits;
	bool seen_dot = false, seen_digit = false;
	for (; p < e; ++p) {
		char c = raw[p];
		if (c >= '0' && c <= '9') {
			(seen_dot ? frac_digits : int_digits).push_back(c);
			seen_digit = true;
		} else if (c == '.' && !seen_dot) {
			seen_dot = true;
		} else {
			break;
		}
	}
	if (!seen_digit) return false;
	long long exp = 0;
	if (p < e && (raw[p] == 'e' || raw[p] == 'E')) {
		p++;
		bool eneg = false;
		if (p < e && (raw[p] == '+' || raw[p] == '-')) {
			eneg = raw[p] == '-';
			p++;
		}
		if (p >= e) return false;
		long long ev = 0;
		for (; p < e; ++p) {
			char c = raw[p];
			if (c < '0' || c > '9') return false;
			ev = ev * 10 + (c - '0');
			if (ev > CQL_DIVIDE_MAX_EXP) return false;
		}
		exp = eneg ? -ev : ev;
	} else if (p != e) {
		return false;
	}
	out.digits = int_digits + frac_digits;
	out.exp10 = exp - static_cast<long long>(frac_digits.size());
	if (out.digits.size() > CQL_DIVIDE_MAX_DIGITS) return false;
	size_t nz = out.digits.find_first_not_of('0');
	if (nz == std::string::npos) {
		out.digits = "0";
		out.sign = 1;
		return true;
	}
	// Stripping LEADING zeros does not change the value's exponent
	// ("05"e-1 == "5"e-1); only trailing-zero stripping would.
	out.digits = out.digits.substr(nz);
	if (out.exp10 > CQL_DIVIDE_MAX_EXP || out.exp10 < -CQL_DIVIDE_MAX_EXP) return false;
	return true;
}

std::string increment_digit_string(const std::string &s) {
	std::string res = s;
	int carry = 1;
	for (int i = static_cast<int>(res.size()) - 1; i >= 0 && carry; --i) {
		int d = res[i] - '0' + carry;
		carry = d / 10;
		res[i] = static_cast<char>('0' + d % 10);
	}
	if (carry) res.insert(res.begin(), '1');
	return res;
}

} // namespace

Optional<CqlDivideResult> cql_divide_text(const std::string &a, const std::string &b) {
	DecimalParts pa, pb;
	if (!parse_cql_decimal_text(a, pa)) return NullOpt<CqlDivideResult>();
	if (!parse_cql_decimal_text(b, pb)) return NullOpt<CqlDivideResult>();
	if (pb.digits == "0") return NullOpt<CqlDivideResult>();
	if (pa.digits == "0") {
		CqlDivideResult r;
		r.negative = false;
		r.unscaled = std::string(8, '0');
		return Optional<CqlDivideResult>(r);
	}

	// quotient magnitude = (Da / Db) * 10^(ea - eb) as N / D
	long long shift = pa.exp10 - pb.exp10;
	std::string num = pa.digits;
	std::string den = pb.digits;
	if (shift > 0) {
		num.append(static_cast<size_t>(shift), '0');
	} else if (shift < 0) {
		den.append(static_cast<size_t>(-shift), '0');
	}

	// Integer part: schoolbook long division. When num < den the integer
	// part is 0 and the remainder is num itself.
	std::string int_part = "0";
	std::string rem = "0";
	if (compare_digit_strings(num, den) >= 0) {
		int_part.clear();
		for (char cd : num) {
			if (rem == "0") {
				rem = std::string(1, cd);
			} else {
				rem.push_back(cd);
			}
			int qd = 0;
			while (qd < 9 && compare_digit_strings(rem, den) >= 0) {
				rem = subtract_digit_strings(rem, den);
				qd++;
			}
			int_part.push_back(static_cast<char>('0' + qd));
		}
		size_t nz = int_part.find_first_not_of('0');
		int_part = (nz == std::string::npos) ? "0" : int_part.substr(nz);
	} else {
		rem = num;
	}

	// Nine fractional digits: 8 scale digits + the HALF_UP guard digit.
	std::string frac(9, '0');
	for (int k = 0; k < 9; ++k) {
		if (rem == "0") break;
		rem.push_back('0');
		int qd = 0;
		while (qd < 9 && compare_digit_strings(rem, den) >= 0) {
			rem = subtract_digit_strings(rem, den);
			qd++;
		}
		frac[k] = static_cast<char>('0' + qd);
	}

	std::string magnitude = int_part + frac.substr(0, 8);
	if (frac[8] >= '5') {
		magnitude = increment_digit_string(magnitude);
	}

	// Implementation Decimal range: magnitude < 10^28 means at most 28
	// integer digits after rounding (the scale-8 suffix is 8 digits).
	if (magnitude.size() > 8 && magnitude.size() - 8 > 28) {
		return NullOpt<CqlDivideResult>();
	}

	CqlDivideResult r;
	r.negative = (pa.sign != pb.sign);
	r.unscaled = magnitude;
	return Optional<CqlDivideResult>(r);
}

} // namespace cql
